### PR TITLE
Handle regular expressions in EJSON

### PR DESCRIPTION
**Pull request by sebakerckhof**
Fixes: https://github.com/meteor/meteor/issues/5652

Background (kudos to: https://github.com/meteor/meteor/issues/5652#issuecomment-157544912 for pointing in the right direction):
Meteor multiplexes/re-uses duplicate observers. For fast look-up of duplicate observers it creates a hash with some properties of the cursor (e.g. selector, projection, ...).
However, for this hash it simply uses `JSON.stringify` ( https://github.com/meteor/meteor/blob/a4de554a95455e91b5d7e71c4c611c2e78725d5e/packages/mongo/mongo_driver.js#L1221 ) , which can not stringify regular expressions. This leads to a problem where different queries would re-use the same (wrong) observer.

However, this selector has already passed the `Mongo.Collection._rewriteSelector` (https://github.com/meteor/meteor/blob/906aab4447ace6cedf873abe0f3c99ad5539cadc/packages/mongo/collection.js#L369) method, which should replace regular expression objects with the alternative `{ $regex: '', $options: '' }` syntax. The problem is that `Mongo.Collection._rewriteSelector` doesn't do its work thoroughly and misses regular expressions inside `$elemMatch` amongst others.

Btw, since order of JSON doesn't guarantee the order of keys in an object, this is probably not the best way to 'hash' the selector, but that's another issue.

Approaches considered:

1. Fix `_rewriteSelector`. This would mean replace every occurence of a regular expression, even deeply nested once in `$elemMatch` and `$in` (amongst others?)

This poses 2 problems:
- The `$in`-array does not support  `{ $regex: '', $options: '' }` syntax
- `_rewriteSelector` is called multiple times for each query, and `$in` can sometimes contain long lists (at least in my apps) which would be a potential serious performance hit.

2. Just use a replacer function in the `JSON.stringify` method here: https://github.com/meteor/meteor/blob/a4de554a95455e91b5d7e71c4c611c2e78725d5e/packages/mongo/mongo_driver.js#L1221
This could check if a value is a RegExp instance and replace it with something serializable

This is the least obtrusive way (like 1 line of code) to fix the specific problem mentioned in: https://github.com/meteor/meteor/issues/5652
However, that still leaves a buggy `_rewriteSelector`. And my mom always thought me I shouldn't do half-assed work, which is why I went for option 3

3. Add support to EJSON for RegExp stringify/parsing. This solves the problem with both `_rewriteSelector` and with `observeKey`-hash.

This is IMO the most elegant solution, but it's also the most obtrusive. From what I can make out from the comments in the code the only reason `_rewriteSelector` replaced regular expressions is to make them DDP-serializable (although it's not clear why it should be DDP-serializable), which this approach achieves. 
But there might be something crucial I'm missing here. Please review

### CODE REVIEW

**Comment by hwillson** [comment]
Thanks for taking a look at this @sebakerckhof! I agree, option 3 sounds like the best path forward. I've looked over your changes, and so far everything looks great. Could you look into why the tests are failing though? It looks like they're getting hung up in the [async upsert tests ](https://github.com/meteor/meteor/blob/1eb578927e983676f4fed1d7836c8adf8355be20/packages/mongo/mongo_livedata_tests.js#L1846) for some reason. Once the tests are passing I'll do a more thorough review/testing.

**Comment by hwillson** [comment]
@sebakerckhof - why did you close this? These changes looked promising - did you uncover potential problems?

**Comment by sebakerckhof** [comment]
Oops, I must have accidentally clicked the close button.

**Comment by sebakerckhof** [comment]
@benjamn These tests are failing here because not all browsers (and the test environment) support ` RegExp.prototype.flags`.
However, even if I add the `ecmascript` package as a dependency of the `ejson` package, the test keeps failing in those environments. Shouldn't that have added a polyfill for it?

**Comment by abernix** [comment]
Currently, `core-js/es6/regexp` is not included in [`ecmascript-client-runtime`](https://github.com/meteor/meteor/blob/6bde360b9c075f1c78c3850eadbdfa7fe271f396/packages/ecmascript-runtime-client/runtime.js#L34).  Not sure if this was an intentional omission by @benjamn during [the reorganization](https://github.com/meteor/meteor/commit/02cbf94834bac2e01d8e8f2c3bf73a407cc00bcb) but it went from being included to being not included.

**Comment by sebakerckhof** [comment]
@benjamn can you shed your light on this?

**Comment by benjamn** [comment]
@sebakerckhof If you rebase against `devel`, you should get a version `ecmascript-runtime-client` that includes the `core-js/es6/regexp` polyfills.

**Comment by sebakerckhof** [comment]
@benjamn thanks, seems to work now.

@hwillson tests are passing. I had to close & re-open again because travis had a segmentation fault first time. Circle CI has 1 failure (also segmentation fault) but that should be unrelated.

**Comment by hwillson** [comment]
@sebakerckhof Just a quick FYI; I've added a few extra commits here to help fix the recent CircleCI issues we've been running into. I needed to use one of the open PR's to help troubleshoot, and we we're planning on merging this PR as soon as the tests passed, so it was a good fit for the additional changes / testing. Tests are passing now, so we're all set here. Thanks!

File: packages/ejson/ejson.js [hunk 1]
```diff
@@ -94,6 +94,21 @@ var builtinConverters = [
...
+    },
+    matchObject: function (obj) {
+      return obj instanceof RegExp;
+    },
+    toJSONValue: function (regexp) {
+      return { $regexp: regexp.source, $flags: regexp.flags };
+    },
+    fromJSONValue: function (obj) {
+      //replaces duplicate / invalid flags
+      return new RegExp(obj.$regexp, obj.$flags.replace(/[^gimuy]/g,'').replace(/(.)(?=.*\1)/g, ''));
```
Comment (mitar) [inline packages/ejson/ejson.js]: So global flag gets removed?

Comment (sebakerckhof) [inline packages/ejson/ejson.js]: No, mind the '^', which is a negation.

Comment (mitar) [inline packages/ejson/ejson.js]: Ah, true. :-)

**Pull request merged**